modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java

```diff
@@ -163,11 +164,14 @@ private <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, @Nulla
                         .handle((res, ex) -> {
                             if (isOrCausedBy(InternalSchemaVersionMismatchException.class, ex)) {
                                 assert tx == null : "Only for implicit transactions a retry might be requested";
-                                assert previousSchemaVersion == null || !Objects.equals(schemaVersion, previousSchemaVersion)
-                                        : "Same schema version (" + schemaVersion
-                                        + ") on a retry: something is wrong, is this caused by the test setup?";
+                                assertSchemaVersionIncreased(previousSchemaVersion, schemaVersion);
 
                                 // Repeat.
+                                return withSchemaSync(tx, schemaVersion, action);
+                            } else if (tx == null && isOrCausedBy(IncompatibleSchemaException.class, ex)) {
+                                // Schema has changed while we were executing an implicit transaction: let's retry.
+                                assertSchemaVersionIncreased(previousSchemaVersion, schemaVersion);
+
                                 return withSchemaSync(tx, schemaVersion, action);
                             } else {
                                 return ex == null ? completedFuture(res) : CompletableFuture.<T>failedFuture(ex);
```

Comment:
I would move the null check at the very beginning, it's a bit more clear: we first check the normal outcome and then start checking the exceptions